boa_engine/src/builtins/intl/date_time_format.rs

```diff
@@ -114,3 +117,392 @@ impl DateTimeFormat {
+
+    // 8. If dateStyle is not undefined or timeStyle is not undefined, let needDefaults be false.
+    if !date_style.is_undefined() || !time_style.is_undefined() {
+        need_defaults = false;
+    }
+
+    // 9. If required is "date" and timeStyle is not undefined, then
+    if required.eq("date") && !time_style.is_undefined() {
+        // a. Throw a TypeError exception.
+        return context.throw_type_error("'date' is required, but timeStyle was defined");
+    }
+
+    // 10. If required is "time" and dateStyle is not undefined, then
+    if required.eq("time") && !date_style.is_undefined() {
+        // a. Throw a TypeError exception.
+        return context.throw_type_error("'time' is required, but dateStyle was defined");
+    }
+
+    // 11. If needDefaults is true and defaults is either "date" or "all", then
+    if need_defaults && (defaults.eq("date") || defaults.eq("all")) {
+        // a. For each property name prop of « "year", "month", "day" », do
+        let property_names = vec!["year", "month", "day"];
+        // i. Perform ? CreateDataPropertyOrThrow(options, prop, "numeric").
+        for prop_name in property_names {
+            options
+                .create_data_property_or_throw(prop_name, "numeric", context)
+                .expect("CreateDataPropertyOrThrow must not fail");
+        }
+    }
+
+    // 12. If needDefaults is true and defaults is either "time" or "all", then
+    if need_defaults && (defaults.eq("time") || defaults.eq("all")) {
+        // a. For each property name prop of « "hour", "minute", "second" », do
+        let property_names = vec!["hour", "minute", "second"];
+        // i. Perform ? CreateDataPropertyOrThrow(options, prop, "numeric").
+        for prop_name in property_names {
+            options
+                .create_data_property_or_throw(prop_name, "numeric", context)
+                .expect("CreateDataPropertyOrThrow must not fail");
+        }
+    }
+
+    // 13. Return options.
+    Ok(options)
+}
+
+/// `DateTimeRangeFormat` type contains `rangePatterns` record represented as an object and may or
+/// may not contain `rangePatterns12`, thus this field is declared optional.
+#[cfg(test)]
+pub(crate) struct DateTimeRangeFormat {
+    pub(crate) range_patterns: JsObject,
+    pub(crate) range_patterns12: Option<JsObject>,
+}
+
+/// `StylesRecord` type hash maps for `TimeFormat`, `DateFormat`, `DateTimeFormat` and
+/// `DateTimeRangeFormat`. The key of these maps should be one of the following:
+/// "full", "long", "medium" or "short". Map values contain formatting patterns like
+/// "{year}-{month}-{day}"
+#[cfg(test)]
+pub(crate) struct StylesRecord {
+    pub(crate) time_format: FxHashMap<JsString, JsValue>,
+    pub(crate) date_format: FxHashMap<JsString, JsValue>,
+    pub(crate) date_time_format: FxHashMap<JsString, JsValue>,
+    pub(crate) date_time_range_format:
+        FxHashMap<JsString, FxHashMap<JsString, DateTimeRangeFormat>>,
+}
```

Comment:
Just wanted to comment that we don't need to create our own structures for each option, since the `icu` crate already has definitions for every one of them:

https://docs.rs/icu/latest/icu/datetime/enum.DateTimeFormatOptions.html

 That'll also avoid having to convert between types when we use the library in the future

Reply:
@raskad, @jedel1043, if I understand the API correctly, the entire `DateTimeStyleFormat` subroutine can be replaced with:

```
fn date_time_style_format_new(
    date_style: Option<icu::datetime::options::length::Date>,
    time_style: Option<icu::datetime::options::length::Time>,
) -> DateTimeFormatOptions {
    DateTimeFormatOptions::Length(
        length::Bag {
             date: date_style,
             time: time_style,
            ..Default::default()
        }
    )
}
```

Since these options are going to be used to build `DateTimeFormat` [here](https://tc39.es/ecma402/#sec-initializedatetimeformat) at step 44.b.ii.

Will such approach comply to the specification?

Thanks.

Reply:
I couldn't find anything about ecma402 compliance in the icu docs. If you think it could be done like this, I'd say lets try it out. The only way to really know is getting some of the builtin Intl function to work and look at the 262 tests.

But I think it would be very good if you would save your current code on a separate branch in your fork. That way no effort is lost if we notice that we have to do it differently in the future.

Reply:
> The only way to really know is getting some of the builtin Intl function to work and look at the 262 tests.

I see. These structures don't make much sense without the whole picture. I would propose to remove `date_time_style_format` implementation from this PR completely and add it in scope of `InitializeDateTimeFormat` PR, which I planned to file next.

Please let me know your opinion.

Reply:
Of course, if you think it makes more sense that way go ahead!